geo/geomfn: Implement ST_SwapOrdinates

adzimzf opened the pull request:
Currently PostGIS has the ST_SwapOrdinates, but CRDB hasn't.

Fixes #49050.

Release note (sql change): Implemented the geometry based builtins
`ST_SwapOrdinates`.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54123) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54123) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=54123) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54123)
<!-- Reviewable:end -->

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
+					Typ:  types.Geometry,
+				},
+				{
+					Name: "cstring",
```
nit: rename this to `swap_ordinate_string`.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
+			},
+			Info: infoBuilder{
+				info: `Returns a version of the given geometry with given ordinates swapped.
+The ords parameter is a 2-characters string naming the ordinates to swap. Valid names are: x,y,z and m.`,
```
nit: space after commas, i.e. `x, y, z and m`

otan commented on pkg/geo/geomfn/swap_ordinates.go:
```diff
@@ -0,0 +1,77 @@
+		if o > len(src)/l.Stride() {
+			return ordsIndex, fmt.Errorf("geometry does not have an %s ordinate", string(ords[i]))
+		}
+		ordsIndex[i] = o
```
the `m` ordinate can be index 2, depending on `geom.Layout`. 
you need to use `l.MIndex()` and `l.ZIndex()` for this to be correct. you can unit test this with a `geom.T` object as long as you test only ingress and outgress `geom.T` objects.

otan reviewed (commented):
thanks! this is close, a few nits and you should be good to go.

otan commented on pkg/geo/geomfn/swap_ordinates.go:
```diff
@@ -0,0 +1,77 @@
+
+	newT, err := applyOnCoordsForGeomT(t, func(l geom.Layout, dst, src []float64) error {
+		if len(ords) != 2 {
+			return errInvalidOrdinate
```
nit also: prefer `return errors.New(....)` here, so we get the stack trace.

adzimzf commented:
Hi @otan,
I have some updates

otan commented on pkg/geo/geomfn/swap_ordinates_test.go:
```diff
@@ -0,0 +1,220 @@
+
+// Test_getOrdsIndices tests the index position of Z and M
+// since currently the geo.Geometry doesn't support ordinates Z and M
+func Test_getOrdsIndices(t *testing.T) {
```
Rename this to `TestGetOrdsIndicies` - we do not use `_` characters in tests.

otan commented on pkg/geo/geomfn/swap_ordinates_test.go:
```diff
@@ -0,0 +1,220 @@
+}
+
+func TestSwapOrdinates_Collection(t *testing.T) {
+	tests := []struct {
```
we normally don't separate test cases like this. this seems easy to fold with the above test cases (i.e. reuse the same struct)? *geom.Collection can be used as geom.T, so you can just extend the number of inputs.

otan commented on pkg/geo/geomfn/swap_ordinates_test.go:
```diff
@@ -0,0 +1,220 @@
+func TestSwapOrdinates(t *testing.T) {
+	tests := []struct {
+		name      string
+		input1    geom.T
```
can you rename input1 / input2 to `inputGeom` and `inputSwapOrdinates`?

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
+			},
+			Info: infoBuilder{
+				info: `Returns a version of the given geometry with given ordinates swapped.
+The ords parameter is a 2-characters string naming the ordinates to swap. Valid names are: x, y, z and m.`,
```
nit: if we're saying `ords parameter` here, then we should rename the argument in "swap_ordinate_string" to "ords" - or rename `ords` here to `swap_ordinate_string`

otan reviewed (commented):
Thanks for the update! Requesting some minor updates to tests.

Also looks like CI is not passing. Can you run `make buildshort` after changing the info string?

adzimzf committed:
geo/geomfn: Implement ST_SwapOrdinates

Currently PostGIS has the ST_SwapOrdinates, but CRDB hasn't.

Fixes #49050.

Release note (sql change): Implemented the geometry based builtins
`ST_SwapOrdinates`.

adzimzf commented:
Hi @otan,
I have some new updates,

otan commented:
Thanks for doing this!

bors r+
